code_generation/configuration.py

```diff
@@ -0,0 +1,526 @@
+        if samples is None or self.sample in samples:
+            scopes_to_shift: ScopeList = [
+                scope for scope in shift.get_scopes() if scope in self.scopes
+            ]
+            for scope in scopes_to_shift:
+                shift.apply(scope)
+                config_change = shift.get_shift_config(scope)
+                shiftname = shift.shiftname
+                self.shifts[scope][shiftname] = config_change
+
+    def add_modification_rule(self, scopes: ScopeList, rule: ProducerRule) -> None:
+        """
+        Function used to add a rule to the configuration.
+
+        Args:
+            scopes: The scopes to which the rule should be added. This can be a list of scopes or a single scope.
+            rule: The rule to be added. This must be a ProducerRule object.
+
+        Returns:
+            None
+
+        """
+        if not isinstance(rule, ProducerRule):
+            raise TypeError("Rule must be of type ProducerRule")
+        if not isinstance(scopes, list):
+            scopes = [scopes]
+        rule.set_scopes(scopes)
+        rule.set_global_scope(self.global_scope)
+        self.rules.add(rule)
+
+    def resolve_modifiers(self, configuration_dict: Dict[Any, Any]) -> TConfiguration:
+        """
+        Function used to resolve mofifiers used in the configuration. This function is called by the add_config_parameters function.
+
+        Args:
+            configuration_dict: The configuration dictionary to be resolved.
+
+        Returns:
+            dict: The resolved configuration dictionary.
+        """
+        resolved_dict: TConfiguration = {}
+        for key in configuration_dict:
+            resolved_value = None
+            log.debug("Resolved: {}".format(resolved_dict))
+            log.debug("Parameter: {}".format(key))
+            log.debug("Paramters: {}".format(configuration_dict))
+            if isinstance(configuration_dict[key], dict):
+                resolved_dict.update(self.resolve_modifiers(configuration_dict[key]))
+            elif isinstance(configuration_dict[key], SampleModifier):
+                log.debug("Applying Samplemodifier")
+                resolved_value = configuration_dict[key].apply(self.sample)
+            elif isinstance(configuration_dict[key], EraModifier):
+                log.debug("Applying Eramodifier")
+                resolved_value = configuration_dict[key].apply(self.era)
+            else:
+                resolved_value = configuration_dict[key]
+            resolved_dict.update({key: resolved_value})
+        return resolved_dict
+
+    def remove_empty_scopes(self) -> None:
+        """
+        Function used to remove empty scopes from the configuration. This function is called by the optimize function,
+        which should be called after all configuration changes have been made.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        scopes_to_test = [scope for scope in self.scopes]
```

Comment:
Is this necessary since `self.scopes` is already a list
https://github.com/KIT-CMS/CROWN/blob/eff8756bcf070769078e2c7ed6d840eaac37f1e1/code_generation/configuration.py#L90

Reply:
:+1:

Reply:
nvm, i remember why this was in place here, we also delete the scope from self.scopes, which is also the list we use to iterate the scopes if we do not assign it a new list, thats why here, a new list is created. I will add a comment in the code to mention this.